### PR TITLE
8276670: G1: Move and rename G1CardSetFreePool and related classes

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Currently, only the card sets (remembered sets) use G1CardSetFreePool to give back memory to OS.

After JDK-8254739, this memory reclaiming mechanism could be reused by evacuation failure too. This is a preparation change to allow reuse of this code.

I plan to do this in about 3 steps to smooth the review process:
  1. move G1CardSetFreePool and related classes to new file, rename these classes
  2. refactor these classes to support freeing other freelist
  3. some necessary cleanup

This is to simply move and rename G1CardSetFreePool and related classes, as G1CardSetFreePool and related classes are going to be reused outside of the remembered set, they should be factored out and renamed.
Rename from G1CardSetXxx -> G1BufferListXxx.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8276670](https://bugs.openjdk.java.net/browse/JDK-8276670): G1: Move and rename G1CardSetFreePool and related classes

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6289/head:pull/6289` \
`$ git checkout pull/6289`

Update a local copy of the PR: \
`$ git checkout pull/6289` \
`$ git pull https://git.openjdk.java.net/jdk pull/6289/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6289`

View PR using the GUI difftool: \
`$ git pr show -t 6289`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6289.diff">https://git.openjdk.java.net/jdk/pull/6289.diff</a>

</details>

### CODE REVIEW

**Comment by Hamlin-Li** [comment]
Seems windows x64 build failure is not related to this change.

**Comment by Hamlin-Li** [comment]
Is some one available to have a look at this change? Thanks

File: src/hotspot/share/gc/g1/g1BufferListFreeMemoryTask.hpp [hunk 1]
```diff
@@ -22,17 +22,18 @@
...
 #include "gc/g1/g1ServiceThread.hpp"
+#include "gc/g1/g1BufferListFreePool.hpp"
 #include "gc/g1/g1CardSetMemory.hpp"
 #include "gc/g1/heapRegionRemSet.hpp"
 #include "utilities/growableArray.hpp"
 #include "utilities/ticks.hpp"
 
 // Task handling deallocation of free card set memory.
-class G1CardSetFreeMemoryTask : public G1ServiceTask {
+class G1BufferListFreeMemoryTask : public G1ServiceTask {
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BufferListFreeMemoryTask.hpp:36]: Comment needs to be adapted.

File: src/hotspot/share/gc/g1/g1BufferListFreeMemoryTask.hpp [hunk 2]
```diff
@@ -53,10 +54,10 @@ class G1CardSetFreeMemoryTask : public G1ServiceTask {
   State _state;
 
   // Current total card set memory usage.
-  G1CardSetMemoryStats _total_used;
+  G1BufferListMemoryStats _total_used;
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BufferListFreeMemoryTask.hpp:57]: Comment mentions "card set"

File: src/hotspot/share/gc/g1/g1BufferListFreePool.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  assert(last != nullptr, "must be");
+
+  last->set_next(nullptr);
+
+  // Wait for any in-progress pops to avoid ABA for them.
+  GlobalCounter::write_synchronize();
+  _source->bulk_add(*_first, *last, keep_num, keep_size);
+  _first = cur;
+
+  log_trace(gc, task)("Card Set Free Memory: Returned to VM %zu buffers size %zu", keep_num, keep_size);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BufferListFreePool.cpp:114]: "Card Set"

File: src/hotspot/share/gc/g1/g1BufferListFreePool.hpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+
+  void clear();
+
+  uint num_pools() const { return G1CardSetConfiguration::num_mem_object_types(); }
+};
+
+// A set of free lists holding memory buffers for use by G1CardSetAllocators.
+template<MEMFLAGS flag>
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:58]: G1CardSetAllocators

File: src/hotspot/share/gc/g1/g1BufferListFreePool.hpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+class G1BufferListFreePool {
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:59]: I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long).
I think there is already a CR to reconsider naming of all these classes, maybe add `G1SegmentedArrayBufferList` to the ones that need renaming; in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant.
Sorry for not speaking up earlier about this, but it only came to my mind right now.

It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames. Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1BufferListFreePool.hpp (hunk 5 above)
Comment (Hamlin-Li) [reply src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:59]: Thanks Thomas.

> I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long). 

Sure.

> I think there is already a CR to reconsider naming of all these classes, 

Do you mean https://bugs.openjdk.java.net/browse/JDK-8276299 or some other issue?

> maybe add `G1SegmentedArrayBufferList` to the ones that need renaming; 

If you mean 8276299 above, sure I will add this to the renaming list of 8276299.

> in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant. 

Not quite get your point, would you mind to clarify further?

> Sorry for not speaking up earlier about this, but it only came to my mind right now.

It's OK, it's never late to make necessary changes. :)

> It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames. 

Do you mean not send a pr but just send email to hotspot-gc-dev to discuss this idea of renaming?

> Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.

If you mean 8276299 above, then I think these 2 (8276299 and this one) are not quite related to each other, 8276299 is to unify the naming of "buffer", "node" and "element" (especially in card set area currently), but this one is to factor out Factor out G1CardSetFreePool and enable evac failure to reuse the freeing memory logic in current G1CardSetFreePool.
Sure, I will start a discussion for 8276299 before actually send the pr.

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:59]: > Thanks Thomas.
> 
> > I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long).
> 
> Sure.
> 
> > I think there is already a CR to reconsider naming of all these classes,
> 
> Do you mean https://bugs.openjdk.java.net/browse/JDK-8276299 or some other issue?

Yes.

> 
> > maybe add `G1SegmentedArrayBufferList` to the ones that need renaming;
> 
> If you mean 8276299 above, sure I will add this to the renaming list of 8276299.
> 
> > in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant.
> 
> Not quite get your point, would you mind to clarify further?

A segmented array is an array consisting of segments (as opposed to a single huge array of something); a list of buffers, where a "buffer" is nothing but an array (of some element type) is almost the same. One is a set of chunks of elements (to explicitly use other words), the other too.

> > It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames.
> 
> Do you mean not send a pr but just send email to hotspot-gc-dev to discuss this idea of renaming?

Not to discuss the idea of renaming, 8276299 is pretty clear about that, but discuss the actual mappings from old name to new name. Maybe this will wake up more people to the discussion than hiding it in a PR :) 

> 
> > Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.
> 
> If you mean 8276299 above, then I think these 2 (8276299 and this one) are not quite related to each other, 8276299 is to unify the naming of "buffer", "node" and "element" (especially in card set area currently), but this one is to factor out Factor out G1CardSetFreePool and enable evac failure to reuse the freeing memory logic in current G1CardSetFreePool. Sure, I will start a discussion for 8276299 before actually send the pr.

That's true, but we are renaming something that also contains a "Buffer" in it and seemingly related.

So instead of renaming this `G1BufferListFreePool` again to something else in 8276299, maybe we could save the time and immediately change it to the final name now, then in 8276299 do the same for the other classes.

Then 8276299 will hopefully go through more smoothly too if we already decided on the new names beforehand as well.

Comment (Hamlin-Li) [reply src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:59]: Thanks for clarification, I see your point, will initiate an discussion later. :)

Comment (Hamlin-Li) [reply src/hotspot/share/gc/g1/g1BufferListFreePool.hpp:59]: Discussion thread is here: https://mail.openjdk.java.net/pipermail/hotspot-gc-dev/2021-November/037626.html

**Comment by Hamlin-Li** [comment]
kindly reminder ~

File: src/hotspot/share/gc/g1/g1CardSetMemory.hpp [hunk 6]
```diff
@@ -25,9 +25,9 @@
 #ifndef SHARE_GC_G1_G1CARDSETMEMORY_HPP
 #define SHARE_GC_G1_G1CARDSETMEMORY_HPP
 
+#include "gc/g1/g1SegmentedArrayFreePool.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CardSetMemory.hpp:28]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1CardSetMemory.cpp [hunk 7]
```diff
@@ -25,10 +25,9 @@
 #include "precompiled.hpp"
 
 #include "gc/g1/g1CardSetMemory.inline.hpp"
+#include "gc/g1/g1CardSetContainers.inline.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CardSetMemory.cpp:28]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 8]
```diff
@@ -33,7 +33,7 @@
 #include "gc/g1/g1Arguments.hpp"
 #include "gc/g1/g1BarrierSet.hpp"
 #include "gc/g1/g1BatchedTask.hpp"
-#include "gc/g1/g1CardSetFreeMemoryTask.hpp"
+#include "gc/g1/g1SegmentedArrayFreeMemoryTask.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:36]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1CollectedHeap.hpp [hunk 9]
```diff
@@ -27,7 +27,7 @@
 
 #include "gc/g1/g1BarrierSet.hpp"
 #include "gc/g1/g1BiasedArray.hpp"
-#include "gc/g1/g1CardSetFreeMemoryTask.hpp"
+#include "gc/g1/g1SegmentedArrayFreeMemoryTask.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CollectedHeap.hpp:30]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.cpp [hunk 10]
```diff
@@ -23,7 +23,7 @@
  */
 
 #include "precompiled.hpp"
-#include "gc/g1/g1CardSetFreeMemoryTask.hpp"
+#include "gc/g1/g1SegmentedArrayFreeMemoryTask.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.cpp:26]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.hpp [hunk 11]
```diff
@@ -22,17 +22,18 @@
  *
  */
 
-#ifndef SHARE_GC_G1_G1CARDSETFREEMEMORYTASK_HPP
-#define SHARE_GC_G1_G1CARDSETFREEMEMORYTASK_HPP
+#ifndef SHARE_GC_G1_G1SEGMENTEDARRAYFREEMEMORYTASK_HPP
+#define SHARE_GC_G1_G1SEGMENTEDARRAYFREEMEMORYTASK_HPP
 
 #include "gc/g1/g1ServiceThread.hpp"
+#include "gc/g1/g1SegmentedArrayFreePool.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.hpp:29]: Include order, must be sorted alphabetically

File: src/hotspot/share/gc/g1/g1SegmentedArrayFreePool.hpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+ */
+
+#ifndef SHARE_GC_G1_G1SEGMENTEDARRAYFREEPOOL_HPP
+#define SHARE_GC_G1_G1SEGMENTEDARRAYFREEPOOL_HPP
+
+#include "gc/g1/g1CardSet.hpp"
+#include "gc/g1/g1SegmentedArray.hpp"
+#include "utilities/growableArray.hpp"
+
+// Statistics for a fixed set of buffer lists. Contains the number of buffers and memory
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1SegmentedArrayFreePool.hpp:32]: s/fixed set of buffer lists/segmented array/ 
Contains the number of segments and memory.... Afair that's part of the other renaming change coming up.

File: src/hotspot/share/gc/g1/g1SegmentedArrayFreePool.hpp [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+// used for each. Note that statistics are typically not taken atomically so there
+// can be inconsistencies. The user must be prepared for them.
+class G1SegmentedArrayMemoryStats {
+public:
+
+  size_t _num_mem_sizes[G1CardSetConfiguration::num_mem_object_types()];
+  size_t _num_buffers[G1CardSetConfiguration::num_mem_object_types()];
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1SegmentedArrayFreePool.hpp:39]: The renaming of `_num_buffers` to `_num_segments` should be done in the other renaming change (which is fine).

File: src/hotspot/share/gc/g1/g1YoungCollector.hpp [hunk 14]
```diff
@@ -33,7 +33,7 @@
 class WorkerTask;
 class G1Allocator;
 class G1BatchedTask;
-class G1CardSetMemoryStats;
+class G1SegmentedArrayMemoryStats;
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1YoungCollector.hpp:36]: Must be sorted alphabetically.

File: test/hotspot/gtest/gc/g1/test_g1CardSet.cpp [hunk 15]
```diff
@@ -22,13 +22,15 @@
  */
 
 #include "precompiled.hpp"
+#include "gc/g1/g1SegmentedArrayFreePool.hpp"
```
Comment (tschatzl) [inline test/hotspot/gtest/gc/g1/test_g1CardSet.cpp:25]: Must be sorted alphabetically.

**Comment by albertnetymk** [comment]
I think splitting "move" and "rename" into two PRs (the current commits history is a bit hard to follow) will make the review process much easier and the included meaningful changes more discernible.

File: src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.cpp [hunk 16]
```diff
@@ -32,13 +32,13 @@
...
+constexpr const char* G1SegmentedArrayFreeMemoryTask::_state_names[];
 
-const char* G1CardSetFreeMemoryTask::get_state_name(State value) const {
+const char* G1SegmentedArrayFreeMemoryTask::get_state_name(State value) const {
   return _state_names[static_cast<std::underlying_type_t<State>>(value)];
 }
 
-bool G1CardSetFreeMemoryTask::deadline_exceeded(jlong deadline) {
+bool G1SegmentedArrayFreeMemoryTask::deadline_exceeded(jlong deadline) {
   return os::elapsed_counter() >= deadline;
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1SegmentedArrayFreeMemoryTask.cpp:42]: The include for `os` seems to be missing (pre-existing)

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Comment by Hamlin-Li** [comment]
I will put the "move" in a separate PR and later "rename" in another PR (including Thomas's suggestion till now)

**Comment by Hamlin-Li** [comment]
I will close this pr, as #6514 is the newly created one which only "rename" (including Thomas's suggestion in this pr till now)
